### PR TITLE
Add old record to change notifications

### PR DESCRIPTION
**Pull request by edmofro**
The commits, in order, achieve the following:
1. Added two tests to catch edge cases of modifying an entity relation record
2. Modified change notification to include `old_record` and `new_record` instead of just `record` (which was previously old for deletes and new for creates/updates)
3. Update every instance of `record` being used within a change to deal with either `old_record` or `new_record` (or both!) instead (plus deprecated `onChange` in favour of `notifiers`)
4. Get new edge case tests passing by handling old and new records in the hierarchy cacher

### CODE REVIEW

File: packages/database/src/DatabaseModel.js [hunk 1]
```diff
@@ -10,11 +10,6 @@ export class DatabaseModel {
 
   constructor(database) {
     this.database = database;
-    // Add change handler to database if defined, and this is the singleton instance of the model
```
Comment (edmofro) [inline packages/database/src/DatabaseModel.js:13]: We had two patterns achieving the same thing. I've removed `onChange` in favour of `notifiers`

File: packages/database/src/migrations/20201008205444-IncludeOldRecordInChangeNotifications-modifies-schema.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+};
+
+exports.up = function (db) {
+  return db.runSql(`
+    CREATE OR REPLACE FUNCTION public.notification() RETURNS trigger
+      LANGUAGE plpgsql
+    AS $$
+    DECLARE
+    new_json_record JSONB;
+    old_json_record JSONB;
```
Comment (edmofro) [inline packages/database/src/migrations/20201008205444-IncludeOldRecordInChangeNotifications-modifies-schema.js:24]: See the `down` for the old function

File: packages/database/src/modelClasses/Indicator.js [hunk 3]
```diff
@@ -19,21 +19,22 @@ export class IndicatorModel extends DatabaseModel {
...
-const onChangeUpdateDataElement = async ({ type: changeType, record }, models) => {
-  const { code } = record;
-
+const onChangeUpdateDataElement = async (
+  { type: changeType, new_record: newRecord, old_record: oldRecord },
+  models,
+) => {
   switch (changeType) {
     case 'update':
       return models.dataSource.findOrCreate(
```
Comment (edmofro) [inline packages/database/src/modelClasses/Indicator.js:28]: Perhaps I should also add a `delete` of any dataSource representing the old code, if they are different. What do you think @kael89

File: packages/meditrak-server/src/database/models/Question.js [hunk 4]
```diff
@@ -35,14 +35,19 @@ export class QuestionModel extends DatabaseModel {
...
-  const { code, data_source_id: dataSourceId } = record;
-
+const onChangeUpdateDataElement = async (
+  { type: changeType, old_record: oldRecord, new_record: newRecord },
+  models,
+) => {
   switch (changeType) {
-    case 'update':
+    case 'update': {
+      const { code, data_source_id: dataSourceId } = newRecord;
```
Comment (edmofro) [inline packages/meditrak-server/src/database/models/Question.js:44]: Same question as earlier - should we delete the old data source, if the data source id doesn't match the old record?

**Review by kael89 — CHANGES_REQUESTED** [review]
Great work @edmofro! It was probably annoying to have to deal with those generally unrelated changes, when development for this project is almost done. Still, I believe it was worth the effort anyway: we can already see some advantages and opportunities created by old data handling!

I hope this doesn't have any of the following side effects:
* Significant performance impact
* Errors because of big change payload

But even if those occur, we should probably treat them properly than avoid providing old data altogether.

A few details and a couple of questions

File: packages/database/src/tests/cachers/EntityHierarchyCacher.test.js [hunk 5]
```diff
@@ -153,6 +155,38 @@ describe('EntityHierarchyCacher', () => {
...
+      { entity_hierarchy_id: 'hierarchy_ocean_test' },
+    );
+    await assertRelationsMatch(
+      'project_storm_test',
+      HIERARCHY_STORM_AFTER_RELATION_HIERARCHY_CHANGED,
+    );
+  });
+
+  it('deletes a subtree and rebuilds when an entity relation parent_id changes', async () => {
+    // move aba -> aaa to be in the ocean hierarchy
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/tests/cachers/EntityHierarchyCacher.test.js:175]: Oops good catch

File: packages/meditrak-server/src/database/models/Question.js (hunk 4 above)
Comment (edmofro) [reply packages/meditrak-server/src/database/models/Question.js:44]: Ok :+1:

File: packages/meditrak-server/src/externalApiSync/SyncQueueChangesManipulator.js [hunk 6]
```diff
@@ -25,7 +25,14 @@ export class SyncQueueChangesManipulator {
...
     this.getRecordIds(this.getChangesForRecordType(changes, model.databaseType));
 
-  getRecords = changes => changes.map(c => c.record);
+  getRecords = changes =>
+    changes.map(change => {
+      const { old_record: oldRecord, new_record: newRecord, type } = change;
+      if (type === DELETE) {
+        return oldRecord;
+      }
+      return newRecord;
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/meditrak-server/src/externalApiSync/SyncQueueChangesManipulator.js:34]: Haha I'm happy to go with it - though I also don't mind some wasted lines for clarity ;-)

File: packages/meditrak-server/src/database/models/AccessRequest.js [hunk 7]
```diff
@@ -9,17 +9,16 @@ export class AccessRequestModel extends CommonAccessRequestModel {
...
-      approved,
-      user_id: userId,
-      entity_id: entityId,
-      permission_group_id: permissionGroupId,
-    },
-  },
-  models,
-) {
+async function onApproveCreateUserCountryPermission({ type, new_record: newRecord }, models) {
+  if (type === 'delete') return;
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/meditrak-server/src/database/models/AccessRequest.js:13]: Yeah I think that's right - I don't think anyone has deleted an access_request yet!

File: packages/meditrak-server/src/externalApiSync/SyncQueueChangesManipulator.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (kael89) [reply packages/meditrak-server/src/externalApiSync/SyncQueueChangesManipulator.js:34]: Agree, my main drive is not line count but easier reading! Of course, this is subjective

[the remark this answers is not in the document]
Comment (edmofro) [reply packages/meditrak-server/src/externalApiSync/SyncQueueChangesManipulator.js:34]: Yeah I think we disagree on which is easier reading, but I am only mildly in favour of the `if`, and have used the ternary form many other times. I'm going with that because I believe it's the more common in our codebase.

[the remark this answers is not in the document]
Comment (kael89) [reply packages/meditrak-server/src/externalApiSync/SyncQueueChangesManipulator.js:34]: Good to know, I thought that people often don't use `? :` because of how their code involves, not as a conscious readability choice. It is a matter of taste then, feel free to use whatever you prefer and to ignore the change suggested above

File: packages/database/src/migrations/20201008205444-IncludeOldRecordInChangeNotifications-modifies-schema.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+          new_json_record
+        )::text
+    );
+      RETURN NEW;
+    END IF;
+    IF TG_OP = 'DELETE' THEN
+      old_json_record := public.scrub_geo_data(
+        to_jsonb(OLD),
+        TG_TABLE_NAME
+      );
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/migrations/20201008205444-IncludeOldRecordInChangeNotifications-modifies-schema.js:61]: Good idea, I've done this as well as made a few other changes for similar reasons

File: packages/meditrak-server/src/dhis/DhisChangeDetailGenerator.js [hunk 9]
```diff
@@ -94,6 +94,6 @@ export class DhisChangeDetailGenerator extends ChangeDetailGenerator {
       ...surveyResponseDetailsById,
     };
 
-    return updateChanges.map(c => JSON.stringify(detailsByChangeId[c.record.id]));
+    return updateChanges.map(c => JSON.stringify(detailsByChangeId[c.record_id]));
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/meditrak-server/src/dhis/DhisChangeDetailGenerator.js:97]: Each change is of the form
```
{
  record_type: 'answer',
  record_id: 'XXX',
  type: 'update',
  old_record: null,
  new_record: {
    id: 'XXX',
    text: '30',
    type: 'Number',
    question_id: '59085fdfcc42a44705c032c2',
    survey_response_id: '590950c37bb75a4aeb6e098b'
  }
}
```

As you can see, both `record_id` and `new_record.id` are the same. In the past, when we just had a single `record`, the `record.id` was the old record's id for deletes, and the new record's id for creates and updates. So, I could have updated this spot to `c.new_record.id`, but I felt it was more succinct to just change to `c.record_id`, which holds the same value. This is also consistent with the existing way it's done in many other places.

File: packages/database/src/cachers/EntityHierarchyCacher.js [hunk 10]
```diff
@@ -37,12 +37,25 @@ export class EntityHierarchyCacher {
...
+    // newRecord will exist for create or update changes
+    if (newRecord) {
+      const { entity_hierarchy_id: newHierarchyId, parent_id: newParentId } = newRecord;
+      await this.deleteSubtree(newHierarchyId, newParentId);
+      this.scheduleHierarchyForRebuild(newHierarchyId);
+    }
+
+    // at least one rebuild will have been scheduled
+    // return the promise for all currently scheduled rebuilds
+    return this.scheduledRebuildPromise;
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/cachers/EntityHierarchyCacher.js:58]: Yeah that's the better way actually - this is coupling the batching logic too tightly to this method. Have changed it to match.

File: packages/database/src/cachers/EntityHierarchyCacher.js [hunk 11]
```diff
@@ -37,12 +37,25 @@ export class EntityHierarchyCacher {
...
-    const { entity_hierarchy_id: hierarchyId, parent_id: parentId } = record;
-    await this.deleteSubtree(hierarchyId, parentId);
-    return this.scheduleHierarchyForRebuild(hierarchyId);
+  handleEntityRelationChange = async ({ old_record: oldRecord, new_record: newRecord }) => {
+    // oldRecord will exist for update or delete changes
+    // delete subtree from old parent_id onwards within the associated hierarchy
+    if (oldRecord) {
+      const { entity_hierarchy_id: oldHierarchyId, parent_id: oldParentId } = oldRecord;
+      await this.deleteSubtree(oldHierarchyId, oldParentId);
+      this.scheduleHierarchyForRebuild(oldHierarchyId);
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/cachers/EntityHierarchyCacher.js:46]: Done

File: packages/database/src/DatabaseChangeChannel.js [hunk 12]
```diff
@@ -34,7 +34,8 @@ export class DatabaseChangeChannel extends PGPubSub {
         type: 'update',
         record_type: recordType,
         handler_key: specificHandlerKey,
-        record,
+        old_record: record,
+        new_record: record,
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/DatabaseChangeChannel.js:38]: Yeah, it is a bit weird, but it does serve the practical purpose of saying "mark these records as needing to be handled as though they were updated, even though nothing has changed, because some dependency _has_ been updated, and that needs to be acknowledged by the system". E.g., when a survey response's date changes, we publish changes for all the answers inside, even though nothing about the answer itself has changed.

Because this is pretending to be an 'update', it does need both old_record and new_record to be present, and it's valid that they're both the same

File: packages/meditrak-server/src/tests/dhis/DhisChangeDetailGenerator.test.js [hunk 13]
```diff
@@ -13,13 +13,20 @@ import {
...
+  return {
+    record_type: type,
+    new_record: {
+      id: recordId,
+      ...record,
+    },
+    old_record: {
+      id: recordId,
+      ...record,
+    },
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/meditrak-server/src/tests/dhis/DhisChangeDetailGenerator.test.js:28]: While the DhisChangeDetailGenerator _does_ respond to the `publishRecordUpdates` fairly often, I've taken your suggestion and made `old_record` null, so that these are testing the situation of a newly created record.

File: packages/database/src/tests/cachers/EntityHierarchyCacher.fixtures.js [hunk 14]
```diff
@@ -289,6 +289,64 @@ export const HIERARCHY_STORM_AFTER_PARENT_ID_CHANGES = expandEntityRelations([
...
+]);
+
+// test changes the parent_id of the aba -> aaa entity to abb
+// storm
+//      a
+//      aa
+//      ab
+//   aba  abb
+//      aaa aab
+export const HIERARCHY_STORM_AFTER_RELATION_PARENT_ID_CHANGED = expandEntityRelations([
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/tests/cachers/EntityHierarchyCacher.fixtures.js]: Added

File: packages/database/src/migrations/20201008205444-IncludeOldRecordInChangeNotifications-modifies-schema.js [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+ */
+exports.setup = function (options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+exports.up = function (db) {
+  return db.runSql(`
+    CREATE OR REPLACE FUNCTION public.notification() RETURNS trigger
```
Comment (kael89) [inline packages/database/src/migrations/20201008205444-IncludeOldRecordInChangeNotifications-modifies-schema.js:19]: This function is very readable now!

File: packages/database/src/DatabaseChangeChannel.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (kael89) [reply packages/database/src/DatabaseChangeChannel.js:38]: Sounds good, thanks for the explanation!

File: packages/meditrak-server/src/dhis/DhisChangeDetailGenerator.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (kael89) [reply packages/meditrak-server/src/dhis/DhisChangeDetailGenerator.js:97]: Sounds good, thanks!

**Review by kael89 — APPROVED** [review]
Great work! I believe we're done (for now)! 🎉🥳

**Comment by edmofro** [comment]
Following up on the performance impact: I've been smoke testing with some timing tests on actions that create a lot of changes, such as submitting 250 survey responses in one hit, or editing a lot of existing survey responses, modifying users, etc., and the new version is always as good or better than the old. Seems adding the old record hasn't hurt us, particularly in the context of the other performance improvements.

We are twice as likely to go over the payload size as before, so we'll have to watch out for that, but we've only hit that in the past with crazy big things like geojson or image data.
